controlplane/src/core/controllers/scim.ts

```diff
@@ -0,0 +1,408 @@
+        displayName,
+        locale,
+        externalId,
+        active,
+        groups,
+        meta: {
+          resourceType: 'User',
+        },
+      });
+    },
+  );
+
+  // update a user
+  fastify.put(
+    '/Users/:userID',
+    async (
+      req: FastifyRequest<{
+        Params: { userID: string };
+        Body: {
+          schemas: string[];
+          id: string;
+          userName: string;
+          name: { givenName?: string; familyName?: string; middleName?: string };
+          emails: { primary: boolean; value: string; type: string }[];
+          password: string;
+          groups: string[];
+          active: boolean;
+        };
+      }>,
+      res,
+    ) => {
+      const authContext = req.authContext;
+
+      const userID = req.params.userID;
+      const { name, emails, password, groups, active } = req.body;
+
+      const user = await opts.organizationRepository.getOrganizationMember({
+        organizationID: authContext.organizationId,
+        userID,
+      });
+
+      if (!user) {
+        return res.code(404).send({
+          schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+          detail: 'User not found',
+          status: 404,
+        });
+      }
+
+      const keycloakUsers = await opts.keycloakClient.client.users.find({
+        realm: opts.keycloakRealm,
+        email: user.email,
+      });
+      if (keycloakUsers.length === 0) {
+        return res.code(404).send({
+          schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+          detail: 'User not found',
+          status: 404,
+        });
+      }
+
+      await opts.keycloakClient.updateKeycloakUser({
+        id: userID,
+        enabled: active,
+        firstName: name.givenName,
+        lastName: name.familyName,
+        realm: opts.keycloakRealm,
+        groups,
+        password,
+      });
+
+      return res.code(200).send({
+        schemas: ['urn:ietf:params:scim:schemas:core:2.0:User'],
+        id: userID,
+        userName: user.email,
+        name,
+        emails,
+        active,
+        groups,
+        meta: {
+          resourceType: 'User',
+        },
+      });
+    },
+  );
+
+  // remove the user from the org
+  fastify.patch(
+    '/Users/:userID',
+    async (
+      req: FastifyRequest<{
+        Params: { userID: string };
+        Body: { schemas: string[]; Operations: { op: string; value: { active: boolean } }[] };
+      }>,
+      res,
+    ) => {
+      const authContext = req.authContext;
+
+      const userID = req.params.userID;
+
+      const orgMember = await opts.organizationRepository.getOrganizationMember({
+        organizationID: authContext.organizationId,
+        userID,
+      });
+      if (!orgMember) {
+        return res.code(404).send({
+          schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+          detail: 'User not found',
+          status: 404,
+        });
+      }
+
+      const active = req.body.Operations[0].value.active;
```

Comment:
I''d check if operations are available and don't we need to check for the `op`?

Reply:
we handle it only for activation and deactivation, op will always be 'replace'
Done